README.md
# ambimax/magento2-deployscripts

Bash scripts for packaging Magento2 projects and deploying it as releases on any server.

- [ambimax/magento2-deployscripts](#ambimaxmagento2-deployscripts)
  - [Deployment Workflow](#deployment-workflow)
  - [Quick Usage](#quick-usage)
    - [Package](#package)
    - [Deploy](#deploy)
    - [Install](#install)
    - [Cleanup](#cleanup)
  - [Project setup](#project-setup)
    - [Setup installation](#setup-installation)
    - [Package](#package-1)
    - [Deploy](#deploy-1)
    - [Install](#install-1)
  - [Hooks](#hooks)
    - [Sample hook](#sample-hook)
  - [Functions available within hooks](#functions-available-within-hooks)
    - [error_exit](#error_exit)
    - [info](#info)
    - [symlinkSharedDirectory](#symlinkshareddirectory)
    - [waitFor](#waitfor)
  - [Cleanup](#cleanup-1)
  - [References](#references)
  - [License](#license)
  - [Author](#author)

## Deployment Workflow

![Illustration of Magento deployment workflow](docs/deployment-workflow.png)

**Before Deployment**

1. Package project to project.tar.gz with `package.sh`
2. Upload project.tar.gz to remote location like S3

**On Deployment**

1. `deploy.sh` is triggered on deploy server
2. Artefact is downloaded and extracted to releases/build_2020010202002
3. `vendor/bin/install.sh` in releases/build_2020010202002 is triggered
4. Hooks in releases/build_2020010202002/deploy/${ENVIRONMENT}/ are triggered
5. All good -> Symlink releases/current to extracted folder releases/build_2020010202002

## Quick Usage

### Package

```bash
./scripts/package.sh \
  --source-dir "${PWD}/tests/workspace/releases/build_dummy/" \
  --build 99 \
  --git-revision 37ed7a1 \
  --filename project.tar.gz
```

### Deploy

```bash
./scripts/deploy.sh \
  --package-url "tests/workspace/artifacts/project.tar.gz" \
  --target-dir /workspace/tests/workspace/ \
  --environment staging
```

Triggers composer install (vendor/bin/install.sh)

_Hint: env variables and additional parameters will be passed to vendor/bin/install.sh_

```bash
# Pass parameters to vendor/bin/install.sh script

./scripts/deploy.sh \
  --package-url "artifacts/project.tar.gz" \
  --target-dir "/var/www" \
  --environment staging \
  --shared-dir "/var/shared" \
  --skip-systemstorage-import
```

### Install

```bash
./scripts/install.sh \
  --project-root "${PWD}/tests/workspace/releases/build_dummy" \
  --environment staging
```

### Cleanup

```bash
./scripts/cleanup.sh \
  --releases-dir "${PWD}/tests/workspace/releases/" \
  --releases-to-keep 5
```

## Project setup

### Setup installation

Add magento2-deployscripts to Magento 2 project

```bash
composer require ambimax/magento2-deployscripts=^1.0.0
```

Add configure hooks to your project (see Hooks)

```bash
mkdir -p deploy/{production,staging}/;
{ echo '#!/usr/bin/env bash'; echo "\n\necho 'hook:pre-install:triggered'"; } > deploy/{production,staging}/pre-install.sh;
```

When project is ready for deployment it must be packaged

### Package

| Argument                     | Description                                          |
| ---------------------------- | ---------------------------------------------------- |
| -f \| --filename             | Filename (i.e. projectA.tar.gz)                      |
| -s \| --source-dir           | Source folder /project_root                          |
| -t \| --target-dir           | Target folder (i.e. artifacts/)                      |
| -b \| --build                | Build number                                         |
| -g \|--git-revision          | GIT revision (i.e. 37ed7a1)                          |
| --exclude-file               | File for separating package / extra package files    |
|                              |                                                      |
| --dump-config                | Dump config before packaging                         |
| --skip-di-compile            | Skip comiling dependency injections before packaging |
| --skip-static-content-deploy | Skip generating static content before packaging      |
| --skip-extra-package         | Skip generating extra package                        |
| --ignore-exclude-file        | Ignore exclude file                                  |
| --save-filelist              | Save filenames into text file                        |

```bash
./scripts/package.sh \
  --source-dir "${PWD}/tests/workspace/releases/build_dummy/" \
  --build 99 \
  --git-revision 37ed7a1 \
  --filename project.tar.gz
```

Files are saved to `artifacts/` directory and can be uploaded either to the server or any other location

```bash
# Upload to s3 storage
aws s3 cp artifacts/project.tar.gz s3://bucket/builds/$(BUILD_NUMBER)/project.tar.gz
aws s3 cp artifacts/project.extra.tar.gz s3://bucket/builds/$(BUILD_NUMBER)/project.extra.tar.gz
aws s3 cp artifacts/MD5SUMS s3://bucket/builds/$(BUILD_NUMBER)/MD5SUMS

# or deployment server
scp artifacts/MD5SUMS \
		artifacts/project.tar.gz \
		artifacts/project.extra.tar.gz \
		user@example.com:/home/project/builds/$(BUILD_NUMBER)/
```

Separation of files in package.tar.gz and package.extra.tar.gz is configured by an exclude file:

```txt
.gitignore
.git
dev/
test/
LICENSE.txt
```

By default this file is expected at `deploy/tar_excludes.txt` - otherwise the location must be defined with `--exclude-file /deploy/file.txt`

To package all files into a single artifact use `--ignore-exclude-file` parameter.

If packages are available it can be deployed

### Deploy

| Argument                          | Description                                                                                   |
| --------------------------------- | --------------------------------------------------------------------------------------------- |
| -e \| --environment               | Environment (i.e. staging, production)                                                        |
| -p \| --package-url               | Package url (https, S3 or local file)                                                         |
| -t \| --target-dir                | Target dir (root of releases/ and current/)                                                   |
| --install-extra-package           | Also download and install .extra.tar.gz package                                               |
|                                   |                                                                                               |
| --aws-profile                     | AWS profile name                                                                              |
| --wget-args                       | Wget arguments like --wget-args "--user=USERNAME --password=PASSWORD"                         |
|                                   |                                                                                               |
| _Pass arguments to deployscript:_ |                                                                                               |
| -p \|--project-root               | Project dir - will be set by CURRENT_RELEASE_DIR (i.e. <targetDir>/releases/build_2020202020) |
| -s \|--shared-dir                 | Shared dir (root of pub/media/ and var/log/)                                                  |
| --skip-systemstorage-import       | Skip systemstorage import                                                                     |

**Pass arguments to deployscript**

_Environment variables and unused parameters will be send to vendor/bin/install.sh_

```bash
# From s3 storage
./scripts/deploy.sh \
  --package-url "s3://bucket/builds/$(BUILD_NUMBER)/project.tar.gz" \
  --target-dir /var/www/staging \
  --environment staging

# On same server
./scripts/deploy.sh \
  --package-url "/home/project/builds/$(BUILD_NUMBER)/project.tar.gz" \
  --target-dir /var/www/staging \
  --environment staging

# With additional parameters for vendor/bin/install.sh
./scripts/deploy.sh \
  --package-url "/home/project/builds/$(BUILD_NUMBER)/project.tar.gz" \
  --target-dir /var/www/staging \
  --environment staging
	--shared-dir "/var/shared" \
	--skip-systemstorage-import
```

### Install

| Argument             | Description                                                          |
| -------------------- | -------------------------------------------------------------------- |
| -e \| --environment  | Environment (i.e. staging, production)                               |
| -r \| --project-root | Project root / extracted release folder /.../releases/build_20210201 |
| -s \|--shared-dir    | Shared folder /.../shared/                                           |

Install script is triggered during deployment and environment variables and unused arguments are passed.

## Hooks

The following hooks are triggered during install.sh

| #   | Hook                 | Location                                   |
| --- | -------------------- | ------------------------------------------ |
| 1   | pre                  | deploy/${ENVIRONMENT}/pre.sh               |
| 2   | defaultvalidation \* | deploy/${ENVIRONMENT}/defaultvalidation.sh |
| 2   | pre-install          | deploy/${ENVIRONMENT}/pre-install.sh       |
| 2   | install \*           | deploy/${ENVIRONMENT}/install.sh           |
| 2   | post-install         | deploy/${ENVIRONMENT}/post-install.sh      |
| 2   | cleanup              | deploy/${ENVIRONMENT}/cleanup.sh           |

\*) Replaces default behaviour

### Sample hook

```bash
#!/usr/bin/env bash
# file: deploy/production/pre-install.sh

echo "hook:pre-install:triggered"

echo "Is production environment" || error_exit "Command not working"
```

## Functions available within hooks

### error_exit

Prints colored error message and exits with 1

```bash
cd bin/ || error_exit "Error Message"
```

### info

Prints colored headline

```bash
info "Verify checksums"

# verify checksums...
```

### symlinkSharedDirectory

Validates source and creates a symlink from source to destination. If destination exists, it will be removed.

```bash
# Create symlink from release/build_2020202020/pub/media to shared/pub/media
symlinkSharedDirectory "pub/media"

# Create symlink from release/build_2020202020/generated to shared/generated
symlinkSharedDirectory "generated"

# Create symlink from release/build_2020202020/var/log to shared/var/log
symlinkSharedDirectory "var/log"
```

### waitFor

Wait for a command or connection to succeed

| Parameter           | Description  |
| ------------------- | ------------ |
| --command <command> | Set command  |
| --host <host>       | Set hostname |
| --port <port>       | Set port     |
| --timeout <seconds> | Set port     |

```bash
# Test command
waitFor --command "/script.sh" --timeout 5

# Test tcp port
waitFor --host www.google.com --port 443 --timeout 1
```

## Cleanup

Remove old deployments

| Argument                 | Description                        |
| ------------------------ | ---------------------------------- |
| -r \| --releases-dir     | Releases root folder /../releases/ |
| -n \| --releases-to-keep | Number of builds to keep           |
| --dry-run                | Enable dry run mode                |

```bash
# Remove old deployments except for last 5
./scripts/cleanup.sh \
  --releases-dir "${PWD}/tests/workspace/releases/" \
  --releases-to-keep 5
```

## References

- [Magento 2.4 Technical Details](https://devdocs.magento.com/guides/v2.4/config-guide/deployment/pipeline/technical-details.html)

## License

MIT

## Author

- [Tobias Schifftner](https://www.twitter.com/tschifftner), [Ambimax GmbH](https://www.ambimax.de)


CHANGELOG.md
# [2.0.0](https://github.com/ambimax/magento2-deployscripts/compare/v1.0.0...v2.0.0) (2021-03-11)


* refactor!: dump config is not default ([335a6df](https://github.com/ambimax/magento2-deployscripts/commit/335a6df12c3d069c496bc5d0fac68c5c98f7a34a))


### BREAKING CHANGES

* --skip-dump-config is replaced by --dump-config
feat: added --exclude-file and --ignore-exclude-file parameter

# 1.0.0 (2021-03-09)


### Features

* added cleanup script ([6bfb85d](https://github.com/ambimax/magento2-deployscripts/commit/6bfb85df3d54cb3cccbda7931b9ae844a08e38de))
* added composer.json ([f8b149f](https://github.com/ambimax/magento2-deployscripts/commit/f8b149ffc4fb77e0d344b084dcc4ee20d88340e7))
* added default env, added waitFor helper ([8b8da62](https://github.com/ambimax/magento2-deployscripts/commit/8b8da628e2ce44a99a44420a9c5fe1257b8b9845))
* added scripts for package, install, deploy ([f2fafe2](https://github.com/ambimax/magento2-deployscripts/commit/f2fafe24ad88539e203247a7ac931cf153df0200))
* cleanup and added documentation ([7b00797](https://github.com/ambimax/magento2-deployscripts/commit/7b007978cbef5bd242cfddf21eba92ac98892499))


tests/helper.bash
#!/usr/bin/env bash

# Set PATH variable for local or docker
[ -d /scripts ] && PATH="${PATH}:/scripts" || PATH="${PATH}:scripts"

# Variables
export TEST_WORKSPACE="${PWD}/tests/workspace"

# Mocks
function php() {
	echo "command:triggered:php $*"
}

export -f php


tests/workspace/releases/build_dummy/dev/test.php
<?php

echo "Testfile";


scripts/deploy.sh
#!/usr/bin/env bash
# Author Tobias Schifftner, https://www.twitter.com/tschifftner

# exit when any command fails
set -e
set -o pipefail
set -u

# debug
[ -n "${BATS_ROOT:-}" ] && set -x

########################################################################################################################
# Variables
########################################################################################################################

# Colors
export COLOR_RED="\033[0;31m"
export COLOR_YELLOW="\033[0;33m"
export COLOR_GREEN="\033[0;32m"
export COLOR_OCHRE="\033[38;5;95m"
export COLOR_BLUE="\033[0;34m"
export COLOR_WHITE="\033[0;37m"
export COLOR_RESET="\033[0m"

# Backgrounds
export ERROR_COLOR="\e[41;1;37m"
export WARNING_COLOR="\e[41;1;37m"
export SUCCESS_COLOR="\e[30;48;5;82m"

# Defaults
INSTALL_EXTRA_PACKAGE=0
AWS_ARGS=
AWS_PROFILE=
ENVIRONMENT=
TARGET_DIR=
PACKAGE_URL=

########################################################################################################################
# Functions
########################################################################################################################

error_exit() {
	echo -e "\n${ERROR_COLOR} ${1} ${COLOR_RESET}"
    exit 1;
}

warning() {
	echo -e "\n${WARNING_COLOR} ${1} ${COLOR_RESET}"
}

info() {
    echo -e "${COLOR_GREEN}"
    echo -e "########################################################################################################################"
    echo -e " $1"
    echo -e "########################################################################################################################"
    echo -e "${COLOR_RESET}"
}

copyFile() {
    echo "Copying package ${1} to ${2}"
    cp "${1}" "${2}" || error_exit "Cannot copy"
}

downloadFile() {
    echo "Download package ${1} to ${2}"
    wget --auth-no-challenge "${WGET_ARGS}" "${1}" -O "${2}" || error_exit "Error while downloading ${1}"
}

downloadS3File() {
    echo "Download S3 package ${1} to ${2}"
    aws "${AWS_ARGS}" s3 cp "${1}" "${2}" || error_exit "Error while downloading ${1} from S3"
}

function usage {
    echo ""
    echo "Usage:"
    echo "$0 --package-url <packageUrl> --target-dir <targetDir>  --environment <environment> [--aws-profile <profile>] [--install-extra-package]"
    echo ""
    echo "   -e|--environment            Environment (i.e. staging, production)"
    echo "   -p|--package-url            Package url (https, S3 or local file)"
    echo "   -t|--target-dir             Target dir (root of releases/ and current/)"
    echo "   --install-extra-package     Also download and install .extra.tar.gz package"
    echo ""
    echo "   --aws-profile               AWS profile name"
    echo '   --wget-args                 Wget arguments like --wget-args "--user=USERNAME --password=PASSWORD"'
    echo ""
    echo ""
    echo " All environment variables and unused parameters will be send to vendor/bin/install.sh:"
    echo ""
    echo "   -p|--project-root             Project dir - will be set by CURRENT_RELEASE_DIR (i.e. <targetDir>/releases/build_2020202020)"
    echo "   -s|--shared-dir               Shared dir (root of pub/media/ and var/log/)"
    echo "   --skip-systemstorage-import   Skip systemstorage import"
    echo ""
    exit 0
}

POSITIONAL=()
while [[ $# -gt 0 ]]
do
key="$1"

case $key in
    -p|--package-url)
    PACKAGE_URL="$2"
    shift # past argument
    shift # past value
    ;;
    -t|--target-dir)
    TARGET_DIR="${2%/}"
    shift # past argument
    shift # past value
    ;;
    -e|--environment)
    ENVIRONMENT="$2"
    shift # past argument
    shift # past value
    ;;
    --aws-profile)
    AWS_PROFILE="$2"
    shift # past argument
    shift # past value
    ;;
    --install-extra-package)
    INSTALL_EXTRA_PACKAGE=1
    shift # past argument
    ;;
    -h|--help)
    usage
    shift # past argument
    ;;
    *)    # unknown option
    POSITIONAL+=("$1") # save it in an array for later
    shift # past argument
    ;;
esac
done
[ "${#POSITIONAL[@]}" -eq 0 ] || set -- "${POSITIONAL[@]}" # restore positional parameters

# Define folders
RELEASES_DIR="${TARGET_DIR}/releases"
CURRENT_RELEASE_NAME="build_$(date +%Y%m%d%H%M%S)"
CURRENT_RELEASE_DIR="${RELEASES_DIR}/${CURRENT_RELEASE_NAME}"

EXTRAPACKAGE_URL=${PACKAGE_URL/.tar.gz/.extra.tar.gz}

if [ -n "${AWS_PROFILE}" ] ; then
    AWS_ARGS="--profile ${AWS_PROFILE}"
fi

export RELEASES_DIR
export CURRENT_RELEASE_NAME
export CURRENT_RELEASE_DIR
export ENVIRONMENT

echo -e "\n${COLOR_GREEN}■ deploy.sh \nTobias Schifftner (@tschifftner), Ambimax GmbH\ngithub.com/ambimax"

########################################################################################################################
info "Validation"
########################################################################################################################

[ -z "${PACKAGE_URL}" ] && error_exit "Package url not set (-p|--package-url)"
[ -z "${TARGET_DIR}" ] && error_exit "Target directory not set (-t|--target-dir)"
[ -z "${ENVIRONMENT}" ] && error_exit "Environment not set (-e|--environment)"

# Check if releases folder exists
[ -d "${RELEASES_DIR}" ] || error_exit "Releases dir ${RELEASES_DIR} not found"
[ -d "${CURRENT_RELEASE_DIR}" ] && error_exit "Release folder ${CURRENT_RELEASE_DIR} already exists"


########################################################################################################################
# Info
########################################################################################################################

echo
echo "##################################################################################################################"
echo
echo "Package url:            ${PACKAGE_URL}"
echo "Target dir:             ${TARGET_DIR}"
echo "Environment:            ${ENVIRONMENT}"
echo
echo "Release folder:         ${CURRENT_RELEASE_DIR}"
echo
echo "##################################################################################################################"
echo

########################################################################################################################
# Create tmp dir and make sure it's going to be deleted in any case
########################################################################################################################

TMPDIR=$(mktemp -d)
function cleanup {
    echo "Removing temp dir ${TMPDIR}"
    rm -rf "${TMPDIR}"
}

trap cleanup EXIT


########################################################################################################################
info "Retrieving package"
########################################################################################################################

if [ -f "${PACKAGE_URL}" ] ; then
    copyFile "${PACKAGE_URL}" "${TMPDIR}/package.tar.gz"
    [ "${INSTALL_EXTRA_PACKAGE}" == 1 ] && copyFile "${EXTRAPACKAGE_URL}" "${TMPDIR}/package.extra.tar.gz"
elif [[ "${PACKAGE_URL}" =~ ^https?:// ]] ; then
    downloadFile "${PACKAGE_URL}" "${TMPDIR}/package.tar.gz"
    [ "${INSTALL_EXTRA_PACKAGE}" == 1 ] && downloadFile "${EXTRAPACKAGE_URL}" "${TMPDIR}/package.extra.tar.gz"
elif [[ "${PACKAGE_URL}" =~ ^s3:// ]] ; then
    downloadS3File "${PACKAGE_URL}" "${TMPDIR}/package.tar.gz"
    [ "${INSTALL_EXTRA_PACKAGE}" == 1 ] && downloadS3File "${EXTRAPACKAGE_URL}" "${TMPDIR}/package.extra.tar.gz"
else
    error_exit "Package \"${PACKAGE_URL}\" not found"
fi


########################################################################################################################
info "Extract package into ${CURRENT_RELEASE_DIR}"
########################################################################################################################

mkdir "${CURRENT_RELEASE_DIR}" || error_exit "Error while creating current release folder"

echo "Extracting base package"
tar xzf "${TMPDIR}/package.tar.gz" -C "${CURRENT_RELEASE_DIR}" || error_exit "Error while extracting base package"

if [ "${INSTALL_EXTRA_PACKAGE}" == 1 ] ; then
    echo "Extracting extra package on top of base package"
    tar xzf "${TMPDIR}/package.extra.tar.gz" -C "${CURRENT_RELEASE_DIR}" || error_exit "Error while extracting extra package"
fi


########################################################################################################################
info "Trigger installation"
########################################################################################################################

COMPOSER_INSTALLER="${CURRENT_RELEASE_DIR}/vendor/bin/install.sh"
if [ -f "${COMPOSER_INSTALLER}" ]; then
    echo "Starting composer installer"
	export PROJECT_ROOT="${CURRENT_RELEASE_DIR}"
	bash "${COMPOSER_INSTALLER}" "$@" || error_exit "Composer installer failed"
fi


########################################################################################################################
info "Update symlink to be live"
########################################################################################################################

echo "Settings current (${RELEASES_DIR}/current) to release folder (${CURRENT_RELEASE_DIR})"
ln -sfn "${CURRENT_RELEASE_DIR}" "${RELEASES_DIR}/current" || error_exit "Error while symlinking 'current' to release folder"

########################################################################################################################
# Finished
########################################################################################################################

echo -e "\n${SUCCESS_COLOR} Successfully deployed! ${COLOR_RESET}\n"


scripts/package.sh
#!/usr/bin/env bash
# Author Tobias Schifftner, https://www.twitter.com/tschifftner

# exit when any command fails
set -e
set -o pipefail
set -u

# debug
[ -n "${BATS_ROOT:-}" ] && set -x

########################################################################################################################
# Variables
########################################################################################################################

# Colors
export COLOR_RED="\033[0;31m"
export COLOR_YELLOW="\033[0;33m"
export COLOR_GREEN="\033[0;32m"
export COLOR_OCHRE="\033[38;5;95m"
export COLOR_BLUE="\033[0;34m"
export COLOR_WHITE="\033[0;37m"
export COLOR_RESET="\033[0m"

# Backgrounds
export ERROR_COLOR="\e[41;1;37m"
export SUCCESS_COLOR="\e[30;48;5;82m"

TMPDIR=$(mktemp -d)

filelist="${TMPDIR}/filelist1.txt"
filelist_extra="${TMPDIR}/filelist2.txt"

FILENAME=
BUILD_NUMBER=
GIT_REVISION=

SOURCE_DIR=$PWD
TARGET_DIR='artifacts'
TAR_EXCLUDE_FILE='deploy/tar_excludes.txt'
IGNORE_EXCLUDE_FILE=false
CONFIG_DUMP=false
SKIP_DI_COMPILE=false
SKIP_STATIC_CONTENT_DEPLOY=false
SKIP_EXTRA_PACKAGE=false
SAVE_FILELIST=false

########################################################################################################################
# Functions
########################################################################################################################

error_exit() {
	echo -e "\n${ERROR_COLOR} ${1} ${COLOR_RESET}"
    exit 1;
}

info() {
    echo -e "${COLOR_GREEN}"
    echo -e "########################################################################################################################"
    echo -e "$1"
    echo -e "########################################################################################################################"
    echo -e "${COLOR_RESET}"
}

function usage {
    echo ""
    echo "Usage:"
    echo "$0 --filename <filename> --build <buildNr>  --source-dir <sourceDir> [--git-revision]"
    echo ""
    echo "   -f|--filename                   Filename (i.e. projectA.tar.gz)"
    echo "   -s|--source-dir                 Source folder /project_root"
    echo "   -t|--target-dir                 Target folder (i.e. artifacts/)"
    echo "   -b|--build                      Build number (i.e. 123)"
    echo "   -g|--git-revision               GIT revision (i.e. 37ed7a1)"
    echo "   --exclude-file                  File for separating package / extra package files"
	echo ""
    echo "   --dump-config                   Dump config before packaging"
    echo "   --skip-di-compile               Skip comiling dependency injections before packaging"
    echo "   --skip-static-content-deploy    Skip generating static content before packaging"
    echo "   --skip-extra-package            Skip generating extra package"
    echo "   --ignore-exclude-file           Ignore exclude file"
    echo "   --save-filelist                 Save filenames into text file"
    echo ""
    exit 0
}

cleanup() {
	if [ -f "${TMPDIR}/env.php" ]; then
		echo "Restore app/etc/env.php"
		mv "${TMPDIR}/env.php" "${SOURCE_DIR}/app/etc/env.php"
	fi

	if [ -f "${SOURCE_DIR}/var/.maintenance.flag" ]; then
		echo "Removing maintenance flag"
		rm "${SOURCE_DIR}/var/.maintenance.flag"
	fi

	if [ -d "${TMPDIR}" ]; then
		echo "Removing temporary files"
		rm -rf "${TMPDIR}"
	fi
}

trap cleanup EXIT
trap cleanup QUIT
trap cleanup INT

POSITIONAL=()
while [[ $# -gt 0 ]]
do
key="$1"

case $key in
    -f|--filename)
    FILENAME="$2"
    shift # past argument
    shift # past value
    ;;
    -b|--build)
    BUILD_NUMBER="${2}"
    shift # past argument
    shift # past value
    ;;
    -g|--git-revision)
    GIT_REVISION="$2"
    shift # past argument
    shift # past value
    ;;
    -s|--source-dir)
    SOURCE_DIR=$(echo "${2}" | sed -e "s/\/*$//")
    shift # past argument
    shift # past value
    ;;
    -t|--target-dir)
    TARGET_DIR=$(echo "${2}" | sed -e "s/\/*$//")
    shift # past argument
    shift # past value
    ;;
    --dump-config)
    CONFIG_DUMP=true
    shift # past argument
    ;;
    --skip-di-compile)
    SKIP_DI_COMPILE=true
    shift # past argument
    ;;
    --skip-static-content-deploy)
    SKIP_STATIC_CONTENT_DEPLOY=true
    shift # past argument
    ;;
    --skip-extra-package)
    SKIP_EXTRA_PACKAGE=true
    shift # past argument
    ;;
    --ignore-exclude-file)
    IGNORE_EXCLUDE_FILE=true
	SKIP_EXTRA_PACKAGE=true
    shift # past argument
    ;;
    --exclude-file)
    TAR_EXCLUDE_FILE="$2"
    shift # past argument
    shift # past value
    ;;
    --save-filelist)
    SAVE_FILELIST=true
    shift # past argument
    ;;
    -h|--help)
    usage
    shift # past argument
    ;;
    *)    # unknown option
    POSITIONAL+=("$1") # save it in an array for later
    shift # past argument
    ;;
esac
done
# [ -n "${POSITIONAL[*]}" ] && set -- "${POSITIONAL[*]}" # restore positional parameters

if [ "${IGNORE_EXCLUDE_FILE}" = "true" ]; then
	TAR_EXCLUDE_FILE="${TMPDIR}/tar_excludes.txt"
	touch "${TAR_EXCLUDE_FILE}"
fi

echo -e "\n${COLOR_GREEN}■ package.sh \nTobias Schifftner (@tschifftner), Ambimax GmbH\ngithub.com/ambimax"

########################################################################################################################
info "Validation"
########################################################################################################################

[ -z "${FILENAME}" ] && error_exit "No file name given (--filename)"
[ -z "${BUILD_NUMBER}" ] && error_exit "No build number given (--build)"

cd "${SOURCE_DIR}" || error_exit "Changing directory failed"

[ -f 'composer.json' ] || error_exit "Could not find composer.json"
[ -f 'pub/index.php' ] || error_exit "Could not find pub/index.php"
[ -f 'app/etc/config.php' ] || error_exit "Could not find app/etc/config.php"
[ -f "${TAR_EXCLUDE_FILE}" ] || error_exit "Could not find exclude list ${TAR_EXCLUDE_FILE}"

########################################################################################################################
info "Prepare for packaging"
########################################################################################################################

# Prepare for production
php bin/magento maintenance:enable

echo "Clean caches"
rm -rf pub/static/* var/cache/* var/view_preprocessed/* var/page_cache/* generated/*

# Prepare configuration
[ "${CONFIG_DUMP}" = "true" ] && php bin/magento app:config:dump --no-interaction

if [ -f "app/etc/env.php" ]; then
	echo "Temporary remove app/etc/env.php"
	mv "app/etc/env.php" "$TMPDIR/env.php"
fi

# Generate files
[ "${SKIP_DI_COMPILE}" = "true" ] || php bin/magento setup:di:compile --no-interaction
[ "${SKIP_STATIC_CONTENT_DEPLOY}" = "true" ] || \
	php bin/magento setup:static-content:deploy --force --strategy=standard --jobs="$(nproc)" --max-execution-time=3600

# Write file: build.txt
echo "${BUILD_NUMBER}" > build.txt

# Write file: version.txt
echo "Build: ${BUILD_NUMBER}" > pub/version.txt
printf "Build time: %s\n" "$(date "+%c")" >> pub/version.txt
[ -n "${GIT_REVISION}" ] && echo "Revision: ${GIT_REVISION}" >> pub/version.txt

# Create package
[ -d "${TARGET_DIR}" ] || mkdir -p "${TARGET_DIR}"

########################################################################################################################
info "Start packaging"
########################################################################################################################

BASEPACKAGE="${TARGET_DIR}/${FILENAME}"
echo "Creating base package '${BASEPACKAGE}'"
tar -vczf "${BASEPACKAGE}" \
    --exclude=./var/log \
    --exclude=./pub/media \
    --exclude=./artifacts \
    --exclude=./app/etc/env.php \
    --exclude="${TARGET_DIR}" \
    --exclude=./tmp \
    --exclude-from="${TAR_EXCLUDE_FILE}" . > "$filelist" || error_exit "Creating archive failed"

# Remove ./ line or all files are ignored
sed -i '/^\.\/$/d' "$filelist"

if [ "${SKIP_EXTRA_PACKAGE}" = "true" ]; then
	echo "Skipping extra package...."
else
	EXTRAPACKAGE=${BASEPACKAGE/.tar.gz/.extra.tar.gz}
	echo "Creating extra package '${EXTRAPACKAGE}' with the remaining files"
	tar -vczf "${EXTRAPACKAGE}" \
		--exclude=./var/log \
		--exclude=./pub/media \
		--exclude=./artifacts \
		--exclude=./app/etc/env.php \
		--exclude="${TARGET_DIR}" \
		--exclude=./tmp \
		--exclude-from="$filelist" . > "$filelist_extra" || error_exit "Creating extra archive failed"
fi

if [ "${SAVE_FILELIST}" = "true" ]; then
	cp "$filelist" artifacts/filelist.txt
	cp "$filelist_extra" artifacts/filelist.extra.txt
fi

cd "${TARGET_DIR}" || error_exit "Cannot enter ${TARGET_DIR}"
md5sum -- *.* > MD5SUMS

########################################################################################################################
info "File hashes"
########################################################################################################################
cat MD5SUMS

echo -e "\n${SUCCESS_COLOR} Successfully packaged! ${COLOR_RESET}\n"

